src/parastell.py

```diff
@@ -1,150 +1,602 @@
+        if self.magnets['meshing']:
+            self.magnet_set.mesh_magnets()
 
-        self.rib = cq.Workplane("XY").spline(self.r_loci).close()
+    def construct_source_mesh(self):
+        """Constructs SourceMesh class object.
+        """
+        self.source_mesh = sm.SourceMesh(
+            self.vmec, self.source['num_s'], self.source['num_theta'],
+            self.source['num_phi'], self.source['tor_ext'], self.logger
+        )
+        self.source_mesh.create_vertices()
+        self.source_mesh.create_mesh()
+        self.source_mesh.write('source_mesh.h5m')
 
-    def vmec2xyz(self, poloidal_offset=0):
-        '''
-        Return an N x 3 NumPy array containing the Cartesian coordinates of the points
-        at this toroidal angle and N different poloidal angles, each offset slightly
+    def export_CAD_geometry(self):
+        """Exports stellarator CAD geometry STEP and/or DAGMC neutronics H5M
+        files according to user-specification.
+        """
+        self._construct_components_dict()
 
-        args:
-            poloidal_offset (double) : some offset to apply to the full set of poloidal
-                angles for evaluating the location of the Cartesian points (default: 0)
-        '''
-        return np.array([self.scale * self.vmec.vmec2xyz(self.s, theta, self.phi) 
-                         for theta in (self.theta + poloidal_offset)])
+        if self.export_dict['_step_export']:
+            self.logger.info('Exporting STEP files...')
+            self._step_export()
 
-    def normalize(vec_list):
-        return np.divide(vec_list.T, np.linalg.norm(vec_list, axis=1).T).T
+        if self.export_dict['h5m_export'] == 'cubit':
+            self.logger.info(
+                'Exporting DAGMC neutronics H5M file via Coreform Cubit...'
+            )
+            self._cubit_export()
 
-    def surf_norm(self):
-        '''
-        Approximate the normal to the curve at each poloidal angle by first approximating
-        the tangent to the curve and then taking the cross-product of that tangent with
-        a vector defined as normal to the plane at this toroidal angle
-        '''
+        if self.export_dict['h5m_export'] == 'cad_to_dagmc':
+            self.logger.info(
+                'Exporting DAGMC neutronics H5M file via CAD-to-DAGMC...'
+            )
+            self._gmsh_export()
 
-        eps = 1e-4
-        next_pt_loci = self.vmec2xyz(eps)
+    def _construct_components_dict(self):
+        """Constructs components dictionary for export routine.
+        """
+        self.components = {}
 
-        tangent = next_pt_loci - self.r_loci
+        for component, (name, layer_data) in zip(
+            self.invessel_build.Components,
+            self.build['radial_build'].items()
+        ):
+            self.components[name] = {}
+            self.components[name]['h5m_tag'] = layer_data['h5m_tag']
+            self.components[name]['solid'] = component
```

Comment:
The method essentially combines the in-vessel component and magnet CAD information into a single dictionary but this data structure definitely isn't necessary for the export procedure. I constructed it this way to make component information (volume IDs and material tags) centrally available for the user after CAD construction, meshing, etc.

I figured if the data is going to be structured this way already, I might as well use it, although I agree it might not be the most streamlined methodology.

We can alternatively have the user access this information from separate `InVesselBuild` and `MagnetSet` class objects on the backend.